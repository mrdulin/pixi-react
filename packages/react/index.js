/* eslint-disable global-require */
if (['test', 'development'].includes(process.env.NODE_ENV))
{
    module.exports = require('./dist/index.cjs-dev');
}
else
{
    module.exports = require('./dist/index.cjs');
}
