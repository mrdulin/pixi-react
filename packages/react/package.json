{
  "name": "@pixi/react",
  "version": "7.1.2",
  "description": "Write PixiJS applications using React declarative style.",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pixijs/pixi-react.git",
    "directory": "packages/react"
  },
  "license": "MIT",
  "sideEffects": false,
  "main": "index.js",
  "module": "module.js",
  "jsnext:main": "module.js",
  "typings": "./index.d.ts",
  "files": [
    "dist/",
    "LICENSE",
    "README.md",
    "index.d.ts",
    "index.js",
    "module.js"
  ],
  "scripts": {
    "prebuild": "npm run clean",
    "build": "npm run build:dev && npm run build:prod",
    "build:dev": "cross-env NODE_ENV=development rollup -c",
    "build:prod": "cross-env NODE_ENV=production rollup -c",
    "clean": "rimraf ./dist",
    "test": "jest stage.spec.js",
    "test:ci": "jest --silent --runInBand",
    "test:types": "tsc --noEmit",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "lodash.isnil": "4.0.0",
    "lodash.times": "4.3.2",
    "performance-now": "2.1.0",
    "prop-types": "^15.8.1",
    "cancelable-promise": "^4.3.1"
  },
  "devDependencies": {
    "@testing-library/react": "^14.2.2",
    "canvas": "^2.11.2",
    "react-test-renderer": "18.2.0"
  },
  "peerDependencies": {
    "pixi.js": ">=8.1.0",
    "prop-types": "^15.8.1",
    "react": ">=17.0.0",
    "react-dom": ">=17.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "6d5b98bc6d7eafd1b0016e60a725b2b6c9f46a33"
}
