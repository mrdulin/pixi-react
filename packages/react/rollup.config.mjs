import copy from 'rollup-plugin-copy';
import {
    getBuildFormat,
    getRollupTSConfig,
    isProductionBuild,
} from '../../shared/getRollupConfig.mjs';
import alias from '@rollup/plugin-alias';

const format = getBuildFormat();
const buildType = isProductionBuild() ? '' : '-dev';

const external = [
    'pixi.js',
    '@pixi/react-invariant',
    'lodash.isnil',
    'lodash.times',
    'prop-types',
    'react',
    'react-dom',
    'react-reconciler',
];

let builds;

const mergeOptions = {
    beforePlugins: [
        alias({
            entries: {
                '@react-spring/animated':
                    '../../shared/react-spring-create-host.js',
            },
        }),
        copy({
            targets: [{ src: 'src/global.d.ts', dest: 'dist/types' }],
        }),
    ],
    external,
};

if (format)
{
    builds = [
        getRollupTSConfig(
            `dist/index.${format}${buildType}.js`,
            format,
            mergeOptions
        ),
    ];
}
else
{
    builds = ['cjs', 'es'].map((format) =>
        getRollupTSConfig(
            `dist/index.${format}${buildType}.js`,
            format,
            mergeOptions
        )
    );
}

export default builds;
