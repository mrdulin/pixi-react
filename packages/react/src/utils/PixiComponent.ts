import { invariant } from './invariant';

import type {
    ComponentsType,
    ComponentType,
    PixiReactMinimalExpandoContainer,
    PropsType,
} from '../types';

export const COMPONENTS: ComponentsType = {};

/**
 * Create Component
 *
 * @param {string} type
 * @param {Object} lifecycle methods
 */
export function PixiComponent<
    P extends PropsType,
    Container extends PixiReactMinimalExpandoContainer,
    Instance extends PixiReactMinimalExpandoContainer
>(type: string, lifecycle: ComponentType<P, Container, Instance>)
{
    invariant(!!type, 'Expect type to be defined, got `%s`', type);
    invariant(
        !COMPONENTS[type],
        'Component `%s` could not be created, a component with that name already exists.',
        type
    );

    COMPONENTS[type] = lifecycle;

    return type as unknown as React.ComponentType<P>;
}
