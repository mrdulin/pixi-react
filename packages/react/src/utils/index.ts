export { hasKey, isArrayWithLength, not } from './fp';
export { invariant } from './invariant';
export { COMPONENTS, PixiComponent } from './PixiComponent';
export { applyDefaultProps, getTextureFromProps, PROPS_DISPLAY_OBJECT, PROPS_RESERVED } from './props';
export { eventHandlers, isPointType, parsePoint, pointsAreEqual, setValue } from './pixi';
