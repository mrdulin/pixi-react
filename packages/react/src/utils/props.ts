import { autoDetectSource, Texture } from 'pixi.js';
import { hasKey, not } from './fp';
import { invariant } from './invariant';
import { eventHandlers, setValue } from './pixi';

import type { Container } from 'pixi.js';
import type {
    ContainerSettableProperty,
    PixiReactContainer,
    PixiReactMinimalExpandoContainer,
    PixiReactTexture,
    PropsType,
} from '../types';

export const CHILDREN = 'children';
/**
 * Reserved props
 *
 * @type {Object}
 */
export const PROPS_RESERVED: Partial<Record<keyof Container, true>> = {
    [CHILDREN]: true,
    parent: true,
    worldTransform: true,
    localColor: true,
    localAlpha: true,
    groupAlpha: true,
    groupColor: true,
    groupColorAlpha: true,
    localBlendMode: true,
    groupBlendMode: true,
    localVisibleRenderable: true,
    groupVisibleRenderable: true,
    renderPipeId: true,
    includeInBuild: true,
    measurable: true,
    isSimple: true,
    updateTick: true,
    localTransform: true,
    relativeGroupTransform: true,
    groupTransform: true,
};

/**
 * Default display object props
 * See https://github.com/michalochman/react-pixi-fiber/blob/a4dbddbef0ffbf0f563c64d30766ea28222a51ea/src/props.js#L9
 *
 * @type {Object}
 */
export const PROPS_CONTAINER: Partial<Record<keyof Container, any>>
    = {
        alpha: 1,
        cursor: null,
        filterArea: null,
        filters: null,
        hitArea: null,
        interactive: false,
        mask: null,
        pivot: 0,
        position: 0,
        renderable: true,
        rotation: 0,
        scale: 1,
        skew: 0,
        visible: true,
        x: 0,
        y: 0,
        boundsArea: null,
        isRenderGroup: false,
        blendMode: null,
        tint: 0xffffff,
        angle: 0,
    };

type TypeValidator = {
    typeofs: string[];
    // eslint-disable-next-line @typescript-eslint/ban-types
    instanceofs: Function[];
};

/**
 * Helper util for fetching the texture from props
 * Can be either texture or image
 *
 * @param {string} elementType
 * @param {PIXI.Container} root
 * @param {object} props
 * @returns {PIXI.Texture|null}
 */
export const getTextureFromProps = <Container extends PixiReactContainer>(
    elementType: string,
    root: Container,
    props: PropsType = {}
): PixiReactTexture =>
{
    if (props.texture)
    {
        invariant(
            props.texture instanceof Texture,
            `${elementType} texture needs to be typeof \`Texture\``
        );

        return props.texture;
    }

    // eslint-disable-next-line consistent-return
    const check = (inType: string, validator: TypeValidator) =>
    {
        if (props.hasOwnProperty(inType))
        {
            const valid
                = validator.typeofs.some((t) => typeof props[inType] === t)
                || validator.instanceofs.some((i) => props[inType] instanceof i);

            invariant(valid, `${elementType} ${inType} prop is invalid`);

            return props[inType];
        }
    };

    // TODO: although source can be a Texture this seems to fail internally within PixiJS
    const result
        = check('image', {
            typeofs: ['string'],
            instanceofs: [HTMLImageElement],
        })
        || check('video', {
            typeofs: ['string'],
            instanceofs: [HTMLVideoElement],
        })
        || check('source', {
            typeofs: ['string', 'number'],
            instanceofs: [
                HTMLImageElement,
                HTMLVideoElement,
                HTMLCanvasElement,
            ],
        });

    invariant(!!result, `${elementType} could not get texture from props`);

    const emitChange = (texture: PixiReactTexture) =>
        requestAnimationFrame(() =>
        {
            texture?.__reactpixi?.root?.emit(`__REACT_PIXI_REQUEST_RENDER__`);
        });

    const texture: PixiReactTexture = Texture.from(
        typeof result === 'string'
            ? result
            : autoDetectSource({ resource: result })
    );

    texture.__reactpixi = { root };
    texture.once('update', emitChange);

    emitChange(texture);

    return texture;
};

const filterProps = not(
    hasKey([...Object.keys(PROPS_RESERVED), ...eventHandlers])
);

/**
 * Apply default props on Display Object instance (which are all components)
 *
 * @param {PIXI.DisplayObject} instance
 * @param {Object} oldProps
 * @param {Object} newProps
 */
export function applyDefaultProps<
    Instance extends PixiReactMinimalExpandoContainer
>(instance: Instance, oldProps: PropsType, newProps: PropsType)
{
    let changed = false;

    // TODO: is it possible to have runtime type checking without importing pixi?
    // invariant(
    //     DisplayObject.prototype.isPrototypeOf(instance),
    //     "instance needs to be typeof `DisplayObject`, got `%s`",
    //     typeof instance
    // );

    // update event handlers
    if (!newProps.ignoreEvents)
    {
        const hasRemoveListener = typeof instance.removeListener === 'function';
        const hasOn = typeof instance.on === 'function';

        for (let i = 0; i < eventHandlers.length; i++)
        {
            const evt = eventHandlers[i];

            if (oldProps[evt] !== newProps[evt])
            {
                changed = true;
                if (typeof oldProps[evt] === 'function' && hasRemoveListener)
                {
                    instance.removeListener(evt, oldProps[evt]);
                }
                if (typeof newProps[evt] === 'function' && hasOn)
                {
                    instance.on(evt, newProps[evt]);
                }
            }
        }
    }

    const newPropKeys = Object.keys(newProps || {});

    // hard overwrite all props? can speed up performance
    if (newProps.overwriteProps)
    {
        for (let i = 0; i < newPropKeys.length; i++)
        {
            const p = newPropKeys[i];

            if (oldProps[p] !== newProps[p])
            {
                changed = true;
                setValue(instance, p, newProps[p]);
            }
        }

        // TODO: previously this returned nothing ie. falsy, should we explicitly return false here instead or was it always
        // a subtle bug, test the intention
        return changed;
    }

    const props = newPropKeys.filter(filterProps);

    for (let i = 0; i < props.length; i++)
    {
        const prop = props[i];
        const value = newProps[prop];

        if (newProps[prop] !== oldProps[prop])
        {
            changed = true;
        }

        if (value !== undefined)
        {
            // set value if defined
            setValue(instance, prop, value);
        }
        else if (prop in PROPS_CONTAINER)
        {
            // is a default value, use that
            console.warn(
                `setting default value: ${prop}, from: ${
                    instance[prop as keyof typeof instance]
                } to: ${value} for`,
                instance
            );
            changed = true;
            setValue(
                instance,
                prop,
                PROPS_CONTAINER[prop as ContainerSettableProperty]
            );
        }
        else
        {
            console.warn(
                `ignoring prop: ${prop}, from ${
                    instance[prop as keyof typeof instance]
                } to ${value} for`,
                instance
            );
        }
    }

    return changed;
}
