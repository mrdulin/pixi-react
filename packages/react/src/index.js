import { TYPES } from './utils/element';

/**
 * -------------------------------------------
 * Public API
 * -------------------------------------------
 */

export * from './exports';

export const BitmapText = TYPES.BitmapText;
export const Container = TYPES.Container;
export const Graphics = TYPES.Graphics;
export const NineSliceSprite = TYPES.NineSliceSprite;
export const Sprite = TYPES.Sprite;
export const AnimatedSprite = TYPES.AnimatedSprite;
export const Text = TYPES.Text;
export const TilingSprite = TYPES.TilingSprite;
export const MeshSimple = TYPES.MeshSimple;
export const MeshRope = TYPES.MeshRope;
