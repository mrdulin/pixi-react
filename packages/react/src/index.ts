/// <reference path="./global.d.ts" />
import {
    AnimatedSprite as AnimatedSpriteComponent,
    BitmapText as BitmapTextComponent,
    Container as ContainerComponent,
    Graphics as GraphicsComponent,
    MeshRope as MeshRopeComponent,
    MeshSimple as MeshSimpleComponent,
    NineSliceSprite as NineSliceSpriteComponent,
    ParticleContainer as ParticleContainerComponent,
    Sprite as SpriteComponent, Text as TextComponent, TilingSprite as TilingSpriteComponent
} from './components';
import { PixiReactFiber } from './reconciler';
import { applyDefaultProps, eventHandlers, PixiComponent } from './utils';

import type {
    ComponentsType,
    ReactAnimatedSpriteComponent,
    ReactBitmapTextComponent,
    ReactContainerComponent,
    ReactGraphicsComponent,
    ReactMeshRopeComponent,
    ReactMeshSimpleComponent,
    ReactNineSliceSpriteComponent,
    ReactSpriteComponent,
    ReactTextComponent,
    ReactTilingSpriteComponent,
} from './types';

/**
 * Available tag types
 *
 * @type {Object}
 */
export const TYPES = {
    BitmapText: 'BitmapText',
    Container: 'Container',
    Graphics: 'Graphics',
    NineSliceSprite: 'NineSliceSprite',
    ParticleContainer: 'ParticleContainer',
    Sprite: 'Sprite',
    AnimatedSprite: 'AnimatedSprite',
    Text: 'Text',
    TilingSprite: 'TilingSprite',
    MeshSimple: 'MeshSimple',
    MeshRope: 'MeshRope',
};

export const BitmapText
    = TYPES.BitmapText as unknown as ReactBitmapTextComponent;
export const Container = TYPES.Container as unknown as ReactContainerComponent;
export const Graphics = TYPES.Graphics as unknown as ReactGraphicsComponent;
export const NineSliceSprite
    = TYPES.NineSliceSprite as unknown as ReactNineSliceSpriteComponent;
export const ParticleContainer
    = TYPES.ParticleContainer as unknown as ReactParticleContainerComponent;
export const Sprite = TYPES.Sprite as unknown as ReactSpriteComponent;
export const AnimatedSprite
    = TYPES.AnimatedSprite as unknown as ReactAnimatedSpriteComponent;
export const Text = TYPES.Text as unknown as ReactTextComponent;
export const TilingSprite
    = TYPES.TilingSprite as unknown as ReactTilingSpriteComponent;
export const MeshSimple
    = TYPES.MeshSimple as unknown as ReactMeshSimpleComponent;
export const MeshRope
    = TYPES.MeshRope as unknown as ReactMeshRopeComponent;

const components: ComponentsType = {
    AnimatedSprite: AnimatedSpriteComponent,
    BitmapText: BitmapTextComponent,
    Container: ContainerComponent,
    Graphics: GraphicsComponent,
    NineSliceSprite: NineSliceSpriteComponent,
    ParticleContainer: ParticleContainerComponent,
    MeshSimple: MeshSimpleComponent,
    MeshRope: MeshRopeComponent,
    Sprite: SpriteComponent,
    Text: TextComponent,
    TilingSprite: TilingSpriteComponent,
};

/**
 * Register components, this is side-effecty :(
 * TODO: Consider using a configuration function like the defunct react-modular version, if even only used internally so side
 * effects can be manually listed in package.json
 */
Object.keys(TYPES).forEach((type) => PixiComponent(type, components[type]));

export { withFilters } from './hoc';
export { useApp, useTick } from './hooks';
export { createRoot, render, roots, unmountComponentAtNode } from './render';
export { Stage } from './stage';
export {
    AppConsumer,
    Context as AppContext,
    AppProvider,
    withPixiApp,
} from './stage/provider';
export { applyDefaultProps, eventHandlers, PixiComponent, PixiReactFiber };
