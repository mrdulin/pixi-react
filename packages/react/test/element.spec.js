import { createElement, PixiComponent, TYPES, TYPES_INJECTED } from 'packages/react/src/utils/PixiComponent';
import {
    AnimatedSprite,
    BitmapText,
    Container,
    Graphics,
    MeshRope,
    MeshSimple,
    NineSliceSprite,
    Point,
    Sprite,
    Text,
    Texture,
    TilingSprite
} from 'pixi.js';
import React from 'react';
import Stage from '../src/stage';
import { desyrel } from './__fixtures__/bitmapfonts';
import { emptyTexture } from './__fixtures__/textures';
import parseBitmapFont from './__utils__/parseBitmapFont';
import { render } from '@testing-library/react';

parseBitmapFont(desyrel);

describe('createElement', () =>
{
    test('types', () =>
    {
        expect(TYPES).toMatchSnapshot();
    });

    test('create Container', () =>
    {
        const element = createElement(TYPES.Container);

        expect(element).toBeInstanceOf(Container);
    });

    test('create Text', () =>
    {
        const element = createElement(TYPES.Text, { text: 'foobar' });

        expect(element).toBeInstanceOf(Text);
    });

    test('create Text as a sprite', () =>
    {
        const element = createElement(TYPES.Text, { text: 'foobar', isSprite: true });

        expect(element).toBeInstanceOf(Sprite);
    });

    test('create Sprite', () =>
    {
        const element = createElement(TYPES.Sprite, { texture: emptyTexture });

        expect(element).toBeInstanceOf(Sprite);
    });

    test('create AnimatedSprite', () =>
    {
        const element = createElement(TYPES.AnimatedSprite, { textures: [emptyTexture] });

        expect(element).toBeInstanceOf(AnimatedSprite);
    });

    test('create BitmapText', () =>
    {
        const element = createElement(TYPES.BitmapText, {
            text: 'foobar',
            style: { fontName: 'Desyrel', fontSize: 35, align: 'left' },
        });

        expect(element).toBeInstanceOf(BitmapText);
    });

    test('create TilingSprite', () =>
    {
        const element = createElement(TYPES.TilingSprite, { texture: emptyTexture });

        expect(element).toBeInstanceOf(TilingSprite);
    });

    test('create Graphics', () =>
    {
        const element = createElement(TYPES.Graphics);

        expect(element).toBeInstanceOf(Graphics);
    });

    test('create NineSliceSprite', () =>
    {
        const element = createElement(TYPES.NineSliceSprite, { texture: emptyTexture });

        expect(element).toBeInstanceOf(NineSliceSprite);
    });

    test('create MeshSimple', () =>
    {
        const element = createElement(TYPES.MeshSimple, { texture: emptyTexture });

        expect(element).toBeInstanceOf(MeshSimple);
    });

    test('create MeshRope', () =>
    {
        const element = createElement(TYPES.MeshRope, {
            texture: emptyTexture,
            points: [new Point(0, 0), new Point(20, 20)],
        });

        expect(element).toBeInstanceOf(MeshRope);
    });

    test('get undefined', () =>
    {
        expect(createElement('INVALID')).toBeUndefined();
    });
});

describe('element.applyProps', () =>
{
    let spy;

    beforeAll(() =>
    {
        spy = jest.spyOn(Texture, 'from').mockReturnValue(emptyTexture);
    });

    afterAll(() =>
    {
        spy.mockRestore();
    });

    test('Sprite.applyProps exists', () =>
    {
        const element = createElement(TYPES.Sprite, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('AnimatedSprite.applyProps with images prop exists', () =>
    {
        const element = createElement(TYPES.AnimatedSprite, { images: ['./image.png'] });

        expect(element).toHaveProperty('applyProps');
        expect(spy).lastCalledWith('./image.png');
    });

    test('AnimatedSprite.applyProps with updated image props', () =>
    {
        const element = createElement(TYPES.AnimatedSprite, { images: ['./image.png'] });
        const changed = element.applyProps(element, { images: ['./image.png'] }, { images: ['./new-image.png'] });

        expect(spy).lastCalledWith('./new-image.png');
    });

    test('AnimatedSprite.applyProps with textures prop exists', () =>
    {
        const element = createElement(TYPES.AnimatedSprite, { textures: [Texture.from('./image.png')] });

        expect(element).toHaveProperty('applyProps');
        expect(spy).lastCalledWith('./image.png');
    });

    test('Sprite.applyProps image', () =>
    {
        const element = createElement(TYPES.Sprite, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeFalsy();
    });

    test('Sprite.applyProps texture', () =>
    {
        const element = createElement(TYPES.Sprite, { texture: emptyTexture });

        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './image.png' });

        expect(changed).toBeTruthy();
    });

    test('TilingSprite.applyProps exists', () =>
    {
        const element = createElement(TYPES.TilingSprite, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('TilingSprite.applyProps image', () =>
    {
        const element = createElement(TYPES.TilingSprite, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(changed).toBeFalsy();
        expect(spy).lastCalledWith('./new-image.png');
    });

    test('TilingSprite.applyProps texture', () =>
    {
        const element = createElement(TYPES.TilingSprite, { texture: emptyTexture });

        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './image.png' });

        expect(changed).toBeTruthy();
    });

    test('TilingSprite.applyProps tilePosition', () =>
    {
        const oldPosition = '1, 2';
        const newPosition = { x: 12, y: 20 };
        const element = createElement(TYPES.TilingSprite, { tilePosition: oldPosition, image: './image.png' });

        const changed = element.applyProps(
            element,
            { tilePosition: oldPosition, image: './image.png' },
            { tilePosition: newPosition, image: './image.png' }
        );

        expect(changed).toBeTruthy();
    });

    test('MeshRope.applyProps exists', () =>
    {
        const element = createElement(TYPES.MeshRope, {
            image: './image.png',
            points: [new Point(0, 0), new Point(20, 20)],
        });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('MeshRope.applyProps image', () =>
    {
        const element = createElement(TYPES.MeshRope, {
            image: './image.png',
            points: [new Point(0, 0), new Point(20, 20)],
        });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(
            element,
            { image: './image.png' },
            {
                image: './new-image.png',
                points: [new Point(0, 0), new Point(20, 20)],
            }
        );

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeTruthy();
    });

    test('NineSliceSprite.applyProps exists', () =>
    {
        const element = createElement(TYPES.NineSliceSprite, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('NineSliceSprite.applyProps image', () =>
    {
        const element = createElement(TYPES.NineSliceSprite, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeFalsy();
    });

    test('NineSliceSprite.applyProps texture', () =>
    {
        const element = createElement(TYPES.NineSliceSprite, { texture: emptyTexture });
        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './new-image.png' });

        expect(changed).toBeTruthy();
    });

    test('MeshSimple.applyProps exists', () =>
    {
        const element = createElement(TYPES.MeshSimple, { image: './image.png' });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toHaveBeenCalledWith('./image.png');
    });

    test('MeshSimple.applyProps image', () =>
    {
        const element = createElement(TYPES.MeshSimple, { image: './image.png' });

        expect(spy).lastCalledWith('./image.png');

        const changed = element.applyProps(element, { image: './image.png' }, { image: './new-image.png' });

        expect(spy).lastCalledWith('./new-image.png');
        expect(changed).toBeFalsy();
    });

    test('MeshSimple.applyProps texture', () =>
    {
        const element = createElement(TYPES.MeshSimple, { texture: emptyTexture });

        const changed = element.applyProps(element, { texture: emptyTexture }, { image: './new-image.png' });

        expect(changed).toBeTruthy();
    });

    test('Graphics.applyProps exists', () =>
    {
        const spy = jest.fn();

        const element = createElement(TYPES.Graphics, { draw: spy });

        expect(element).toHaveProperty('applyProps');
        expect(spy).toBeCalledWith(element);
    });

    test('Graphics.applyProps draw', () =>
    {
        const spy = jest.fn();

        const element = createElement(TYPES.Graphics, { draw: spy });

        expect(spy).toHaveBeenCalledTimes(1);

        const applied = element.applyProps(element, { draw: spy }, { draw: spy });

        expect(spy).toHaveBeenCalledTimes(1);
        expect(applied).toBeFalsy();
    });

    test('Graphics.applyProps draw prevented twice', () =>
    {
        const draw1 = jest.fn();
        const draw2 = jest.fn();
        const props = { draw: draw1 };
        const nextProps = { draw: draw2 };
        let applied = false;
        const element = createElement(TYPES.Graphics, props);

        applied = element.applyProps(element, props, props);
        expect(applied).toBeFalsy();
        applied = element.applyProps(element, props, props);
        expect(applied).toBeFalsy();
        expect(draw1).toHaveBeenCalledTimes(1);
        applied = element.applyProps(element, props, nextProps);
        expect(applied).toBeTruthy();
        applied = element.applyProps(element, nextProps, nextProps);
        expect(applied).toBeFalsy();
        expect(draw2).toHaveBeenCalledTimes(1);
    });
});

describe('PixiComponent', () =>
{
    afterEach(() =>
    {
        Object.keys(TYPES_INJECTED).forEach((k) =>
        {
            delete TYPES_INJECTED[k];
        });
    });

    test('type must be defined', () =>
    {
        expect(() => new PixiComponent(null)).toThrow('Expect type to be defined, got `null`');
    });

    test('cannot override existing component', () =>
    {
        expect(() => new PixiComponent('Text', {})).toThrow(
            'Component `Text` could not be created, already exists in default components.'
        );
    });

    test('inject custom component', () =>
    {
        const lifecycle = { create: (props) => {} };

        // eslint-disable-next-line no-new
        new PixiComponent('Rectangle', lifecycle);
        expect(TYPES_INJECTED).toHaveProperty('Rectangle', lifecycle);
    });

    test('create injected component', () =>
    {
        const scoped = jest.fn();
        const config = {
            destroyChildren: true,
            destroy: true
        };

        const lifecycle = {
            create: jest.fn(() => new Graphics()),
            didMount: jest.fn(),
            willUnmount: jest.fn(),
            applyProps: jest.fn(function applyProps()
            {
                scoped(this);
            }),
            config
        };

        // eslint-disable-next-line no-new
        new PixiComponent('Rectangle', lifecycle);

        const props = { x: 100, y: 200 };
        const element = createElement('Rectangle', props, null);

        expect(element.didMount).toBeDefined();
        expect(element.willUnmount).toBeDefined();
        expect(element.applyProps).toBeDefined();
        expect(element.config).toBe(config);
        expect(element).toBeInstanceOf(Graphics);
        expect(lifecycle.create).toHaveBeenCalledTimes(1);
        expect(lifecycle.create).toHaveBeenCalledWith(props, { root: null });
        expect(lifecycle.applyProps).toHaveBeenCalledTimes(1);
        expect(scoped).toHaveBeenCalledTimes(1);
        expect(scoped).toHaveBeenCalledWith(element);
    });

    test.each`
        destroyChildren
        ${true}
        ${false}
    `('unmount and destroy component when destroyChildren is $destroyChildren', ({ destroyChildren }) =>
    {
        const scoped = jest.fn();
        const config = {
            destroyChildren,
            destroy: true
        };

        const makeLifecycle = (pixiInstance) => ({
            create: jest.fn(() => pixiInstance),
            didMount: jest.fn(),
            willUnmount: jest.fn(),
            applyProps: jest.fn(function applyProps()
            {
                scoped(this);
            }),
            config
        });

        const parentPixiInstance = new Container();
        const childPixiInstance = new Container();

        const parentDestroySpy = jest.spyOn(parentPixiInstance, 'destroy');
        const childDestroySpy = jest.spyOn(childPixiInstance, 'destroy');

        const parentLifecycle = makeLifecycle(parentPixiInstance);
        const childLifecycle = makeLifecycle(childPixiInstance);

        const Parent = PixiComponent('Parent', parentLifecycle);
        const Child = PixiComponent('Child', childLifecycle);

        const { unmount } = render(
            <Stage width={800} height={600}>
                <Parent>
                    <Child />
                </Parent>
            </Stage>
        );

        expect(parentLifecycle.create).toHaveBeenCalledTimes(1);
        expect(parentLifecycle.didMount).toHaveBeenCalledTimes(1);
        expect(parentLifecycle.applyProps).toHaveBeenCalledTimes(1);
        expect(childLifecycle.create).toHaveBeenCalledTimes(1);
        expect(childLifecycle.didMount).toHaveBeenCalledTimes(1);
        expect(childLifecycle.applyProps).toHaveBeenCalledTimes(1);

        unmount();

        expect(parentLifecycle.willUnmount).toHaveBeenCalledTimes(1);
        expect(parentDestroySpy).toHaveBeenCalledTimes(1);
        expect(parentDestroySpy).toHaveBeenCalledWith({
            children: destroyChildren,
            texture: false,
            baseTexture: false
        });

        if (destroyChildren)
        {
            expect(childLifecycle.willUnmount).toHaveBeenCalledTimes(1);
            expect(childDestroySpy).toHaveBeenCalledTimes(1);
            expect(childDestroySpy).toHaveBeenCalledWith({
                children: destroyChildren,
                texture: false,
                baseTexture: false
            });
        }
        else
        {
            expect(childLifecycle.willUnmount).not.toHaveBeenCalled();
            expect(childDestroySpy).not.toHaveBeenCalled();
        }
    });

    test('create injected component without lifecycle methods', () =>
    {
        // eslint-disable-next-line no-new
        new PixiComponent('Rectangle', {
            create: () => new Graphics(),
        });

        const element = createElement('Rectangle');

        expect(element.didMount).toBeUndefined();
        expect(element.willUnmount).toBeUndefined();
        expect(element.applyProps).toBeUndefined();
    });
});
