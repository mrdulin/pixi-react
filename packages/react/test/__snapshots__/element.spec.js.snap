// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createElement types 1`] = `
{
  "AnimatedSprite": "AnimatedSprite",
  "BitmapText": "BitmapText",
  "Container": "Container",
  "Graphics": "Graphics",
  "MeshRope": "MeshRope",
  "NineSliceSprite": "NineSliceSprite",
  "MeshSimple": "MeshSimple",
  "Sprite": "Sprite",
  "Text": "Text",
  "TilingSprite": "TilingSprite",
}
`;
